Unify nature checks and reuse existing constants from core.natures.PDE

HannesWell opened the pull request.

HannesWell commented:
I have now taken this as an opportunity to replace all occurrences of PDE nature and builder literals by the corresponding constants in the `PDE` class and to leveraging the different PDE.hasXNature methods.

Additionally I moved the `isBndProject(IProject)` method from `BndProject` to the PDE class to align it with the other methods.

I also contemplate to rename the `hasXNature()` method to `isXProject()`, which I would consider a more speaking name.
But if these names are changed, we could alternativly consider to do the opposite and move the hasXNature respectivly isXProject methods as well as the constants to the corresponding XProject classes in the same package, just like it used to be for the BndProject class.

What's the opinion of the others? Keep one central class for all nature and builder ids and `hasNature()` or `isProject()` methods? Or should those constants and methods be added to the corresponding project class?
At the moment we have a split between BNDProject and Plugin/Feature/SiteProject and I would like to unify this.

laeubi commented:
I would appreciate if we could delay such cleanups/unification until the pending PRs are merged so they don't get delayed/distorted by rebase conflicts, for me central classes tend to be convoluted by more and more unrelated stuff over time so I think keeping it on the project classes seems good.

laeubi commented on apitools/org.eclipse.pde.api.tools.ui/src/org/eclipse/pde/api/tools/ui/internal/wizards/ApiToolingSetupWizardPage.java:
```diff
@@ -268,11 +269,12 @@ private IProject[] getInputProjects() {
 			return false;
 		}
-		return (project.hasNature(JavaCore.NATURE_ID) && project.hasNature("org.eclipse.pde.PluginNature")) //$NON-NLS-1$
+		return project.hasNature(JavaCore.NATURE_ID) && PDE.hasPluginNature(project)
```
I think we should have also a isJavaPluginNature(...) that checks for java+plugin nature).

merks replied:
I assume you mean **has**JavaPluginNature...

HannesWell replied:
I had a look at this but didn't found anymore case where this would useful.
For now I just added a `PluginProject.isJavaProject()` method that can be used in independently of `PluginProject.isPluginProject()`.
If later more use cases are discovered for a combined method we can still add it.

laeubi replied:
At laest one can create a Plugin Project without a java nature in the wizzard (even though it is seldom useful for me...)

HannesWell replied:
Yes, but I did the changes in this PR purely from a code point of view.

laeubi commented:
@HannesWell still relevant?

HannesWell commented:
> I also contemplate to rename the `hasXNature()` method to `isXProject()`, which I would consider a more speaking name.
> But if these names are changed, we could alternativly consider to do the opposite and move the hasXNature respectivly isXProject methods as well as the constants to the corresponding XProject classes in the same package, just like it used to be for the BndProject class.

Implemented that now which makes the PDE class obsolete since all methods and constants have been moved to the corresponding `<Type>Project` classes. Furthermore I renamed `BaseProject` to `PDEProject` since fundamental constants and methods reside there now and I think the latter name more expressive.

HannesWell commented:
This should finally work.
@laeubi or @merks do you have any remarks on this one?

laeubi commented on apitools/org.eclipse.pde.api.tools.tests/src/org/eclipse/pde/api/tools/builder/tests/ApiBuilderTest.java:
```diff
@@ -1154,7 +1154,7 @@ private static String toString(List<IProject> projects) throws Exception {
 			contents.append("    is open: " + project.isOpen()); //$NON-NLS-1$
 			contents.append(System.lineSeparator());
-			if (project.hasNature(JavaCore.NATURE_ID) && project.isAccessible()) {
+			if (Util.isJavaProject(project) && project.isAccessible()) {
```
Whats the difference between `Util` and `ProjectUtils`?

I'm not really a fan of Util classes as they are often a sign of bad API and tend to collect all kind of stuff and dependencies, but do we really need two of them?

HannesWell replied:
> Whats the difference between `Util` and `ProjectUtils`?
> 
Which `ProjectUtils` are you referring? `Util` is a utility class for API tools that already exists and does much more than just this.

laeubi replied:
> Which ProjectUtils are you referring? 

https://github.com/eclipse-pde/eclipse.pde/pull/941/files#diff-42ce900b7490497be3f213a0f5930372f29db30ff877df84fbfc78245cd7b392

HannesWell replied:
> I'm not really a fan of Util classes as they are often a sign of bad API and tend to collect all kind of stuff and dependencies, but do we really need two of them?

I want to emphasize that all changes in this PR are internal and no API is involved at all.
And yes I tried to reduce the duplications with this PR. I kept that method since `PluginProject` is not an API calls and I didn't want to add yet another friend to its package. But in the end I already added more so another one doesn't matter.

HannesWell commented:
These quality-gates are really pedantic, but in the end that's good. And now finally the build is green. 🎉
Submitting.
